### PR TITLE
Add very basic documentation for interop with Swift/Objective-C

### PR DESCRIPTION
**Pull request by SvyatoslavScherbina**

### CODE REVIEW

**Comment by SvyatoslavScherbina** [comment]
To be merged after #1481

File: OBJC_INTEROP.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
+# _Kotlin/Native_ interoperability with Swift/Objective-C
+
+This documents covers some details of Kotlin/Native interoperability with
+Swift/Objective-C.
+
+## Usage
+
+Kotlin/Native provides bidirectional interoperability with Objective-C.
+Objective-C frameworks and libraries can be used in Kotlin code if
+properly attached to the build (system frameworks are attached by default).
```
Comment (olonho) [inline OBJC_INTEROP.md]: "attached to" -> "integrated with", attachment assumes somewhat asymmetrical relation.

File: OBJC_INTEROP.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+See e.g. "Interop libraries" in
+[Gradle plugin documentation](GRADLE_PLUGIN.md#building-artifacts).
+Swift library can be used in Kotlin code if its API is exported to Objective-C
+with `@objc`.
```
Comment (olonho) [inline OBJC_INTEROP.md]: Maybe explicitly mention "Pure Swift modules are not yet supported"?

File: OBJC_INTEROP.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+Kotlin module can be used in Swift/Objective-C code if compiled into a
+[framework](GRADLE_PLUGIN.md#framework). See [calculator sample](samples/calculator)
+as an example.
+
+## Mappings
+
+The table below shows how Kotlin concepts are mapped to Swift/Objective-C and vice versa.
+
+| Kotlin | Swift | Objective-C | Notes |
```
Comment (olonho) [inline OBJC_INTEROP.md]: +100!

File: OBJC_INTEROP.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+Objective-C classes are imported into Kotlin with their original names.
+Protocols are imported as interfaces with `Protocol` name suffix,
+i.e. `@protocol Foo` -> `interface FooProtocol`.
+These classes and interfaces are placed into package [specified in build configuration](#usage)
+(`platform.*` packages for preconfigured system frameworks).
+
+Names of Kotlin classes and interfaces are prefixed when imported to Swift/Objective-C.
+The prefix is derived from the framework name.
+
+### Top-level functions and properties
```
Comment (olonho) [inline OBJC_INTEROP.md]: Unit conversion paragraph?

File: OBJC_INTEROP.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+Top-level Kotlin functions and properties are accessible as members of a special class.
```
Comment (olonho) [inline OBJC_INTEROP.md]: Example?

File: OBJC_INTEROP.md [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+### Category members
+
+Members of Objective-C categories and Swift extensions are imported to Kotlin
+as extensions. That's why these declarations can't be overridden in Kotlin.
+And extension initializers aren't available as Kotlin constructors.
+
+### Number
+
+`NSNumber` is not automatically translated to Kotlin primitive types
+when used as Swift/Objective-C parameter type or return value.
```
Comment (olonho) [inline OBJC_INTEROP.md]: Maybe some rationale, or not yet implemented motivation?

File: OBJC_INTEROP.md [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+### NSMutableString
+
+`NSMutableString` Objective-C class is not available from Kotlin.
+All instances of `NSMutableString` are copied when passed to Kotlin.
+
+### Mutable collections
+
+Every Kotlin `MutableSet` is `NSMutableSet`, however the opposite is not true.
```
Comment (olonho) [inline OBJC_INTEROP.md]: Please mention explicitly that all other collections, such as Map and List are properly mapped.

File: OBJC_INTEROP.md [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+val string = nsString as String
+val nsNumber = 42 as NSNumber
+```
+
+## Subclassing
+
+Kotlin classes and interfaces can be subclassed by Swift/Objective-C classes
+and protocols.
+Currently a class that adopts Kotlin protocol should inherit `NSObject`
+(either directly or indirectly). Note that all Kotlin classes do inherit `NSObject`.
```
Comment (olonho) [inline OBJC_INTEROP.md]: What do you mean here?

File: OBJC_INTEROP.md [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+Swift/Objective-C classes can be subclassed with Kotlin `final` class.
```
Comment (olonho) [inline OBJC_INTEROP.md]: Please mention, that non-final inheritance and complex object hierarchies are not yet supported.

File: OBJC_INTEROP.md (hunk 2 above)
Comment (SvyatoslavScherbina) [reply OBJC_INTEROP.md]: Instead of the last sentence or after it?

File: OBJC_INTEROP.md (hunk 4 above)
Comment (SvyatoslavScherbina) [reply OBJC_INTEROP.md]: Yes, it makes sense. What do you think, should it be related to function types mapping or to `Unit` type mapping?

File: OBJC_INTEROP.md (hunk 7 above)
Comment (SvyatoslavScherbina) [reply OBJC_INTEROP.md]: Actually while writing these paragraphs I was considering them to be read just as notes to the table, after following the corresponding links.
Now I see that they look very silly when being read as self-contained descriptions. So I suppose some of the paragraphs should be expanded.

File: OBJC_INTEROP.md (hunk 8 above)
Comment (SvyatoslavScherbina) [reply OBJC_INTEROP.md]: All Kotlin classes are visible to Swift/Objective-C as inheriting `NSObject`, so it is OK to adopt Kotlin protocol by a subclass of a Kotlin class.

File: OBJC_INTEROP.md [hunk 10]
```diff
@@ -0,0 +1,124 @@
+# _Kotlin/Native_ interoperability with Swift/Objective-C
+
+This documents covers some details of Kotlin/Native interoperability with
+Swift/Objective-C.
```
Comment (SvyatoslavScherbina) [inline OBJC_INTEROP.md]: Shouldn't we use some common name for Swift and Objective-C?

File: OBJC_INTEROP.md (hunk 2 above)
Comment (olonho) [reply OBJC_INTEROP.md]: After

File: OBJC_INTEROP.md (hunk 4 above)
Comment (olonho) [reply OBJC_INTEROP.md]: Maybe function types.

File: OBJC_INTEROP.md (hunk 8 above)
Comment (olonho) [reply OBJC_INTEROP.md]: Probably say that somehow more clear.

File: OBJC_INTEROP.md (hunk 10 above)
Comment (olonho) [reply OBJC_INTEROP.md]: Good idea, although I've no idea what could it be.

File: OBJC_INTEROP.md (hunk 1 above)
Comment (SvyatoslavScherbina) [reply OBJC_INTEROP.md]: But why this relation should be described as being symmetrical?
I mean, when importing a library to Kotlin code through `cinterop`, we don't import Kotlin code to the library.

We may use replace "attach" by "include" or "import".

File: OBJC_INTEROP.md [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+
+## Mappings
+
+The table below shows how Kotlin concepts are mapped to Swift/Objective-C and vice versa.
+
+| Kotlin | Swift | Objective-C | Notes |
+| ------ | ----- |------------ | ----- |
+| `class` | `class` | `@interface` | [note](#name-translation) |
+| `interface` | `protocol` | `@protocol` | |
+| `constructor` | Initializer | Initializer | |
```
Comment (r4zzz4k) [inline OBJC_INTEROP.md]: I'm not sure if constructors map to initializers in a natural way for Kotlin developers, so there probably should be some note on `initBy(...)` and invocation ordering.

Background:
I'm playing around with Kotlin/Native based iOS app and finding initializers a bit confusing (having no background in Obj-C / Swift).
I'm trying out code-only approach, so I've removed storyboards and invoking this in `AppDelegate`:
```kotlin
val mainViewController = ViewController()
println("mainViewController ready [${mainViewController.hashCode()}]")
```
In my `ViewController` I've added logging to constructors and `initWith...` methods as following:
```kotlin
@ExportObjCClass
class ViewController: UIViewController {

    init {
        println("ViewController [${hashCode()}] ctor: common init")
    }

    constructor(): super() {
        println("ViewController [${hashCode()}] ctor: no-arg")
    }
    constructor(aDecoder: NSCoder): super(aDecoder) {
        println("ViewController [${hashCode()}] ctor: NSCoder")
    }
    constructor(nibNameOrNil: String?, bundle: NSBundle?): super(nibNameOrNil, bundle) {
        println("ViewController [${hashCode()}] ctor: nibNameOrNil")
    }

    override fun init(): UIViewController? {
        println("ViewController [${hashCode()}]: init")
        return initBy(ViewController())
    }
    override fun initWithCoder(aDecoder: NSCoder): UIViewController? {
        println("ViewController [${hashCode()}]: initWithCoder")
        return initBy(ViewController(aDecoder))
    }
    override fun initWithNibName(nibNameOrNil: String?, bundle: NSBundle?): UIViewController {
        println("ViewController [${hashCode()}]: initWithNibName")
        return initBy(ViewController(nibNameOrNil, bundle))
    }
    // ...
}
```

Here's the resulting output:
```
ViewController [1807837959]: initWithNibName
ViewController [1807837959] ctor: common init
ViewController [1807837959] ctor: nibNameOrNil
ViewController [1807837959] ctor: common init
ViewController [1807837959] ctor: no-arg
mainViewController ready [1807837959]
```

So there are several points I'm finding odd, which ask for being covered by interop documentation:
* No-arg constructor leads to `initWithNibName`. This is probably due to [`init(nibName:bundle:)`](https://developer.apple.com/documentation/uikit/uiviewcontroller/1621359-init) being [designated initializer](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/Initialization.html#//apple_ref/doc/uid/TP40014097-CH18-ID217).
* `initWithNibName` is being called before any constructor code. Looks like that's caused by the nature of initializers in Objective-C, so they are treated a bit differently from other methods. I see that `initBy` is intrinsic and [has a special handling](https://github.com/JetBrains/kotlin-native/blob/77139d58f822aaef72498570371c7448426a695d/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt#L1686) when invoked, but it's not documented in any way, so a note here might be helpful. And actual `initWith...` methods appear as usual methods in generated `.kt` stubs, so nothing reveals their real nature.
* Two different constructors being called on one object, with two calls to `init { ... }` block. I guess Objective-C base class is able to invoke another constructor of derived class, and I'm not sure if such behavior can be achieved with Kotlin/JVM, so this also is a surprise point.

Please let me know if I'm doing something wrong or if something should be filled as an issue -- I'll try to provide as much details as possible.

Comment (SvyatoslavScherbina) [reply OBJC_INTEROP.md]: @r4zzz4k overriding of initializers is tricky and considered to be redesigned.
The problem with your example is that you call `super`-constructor that is based on non-designated initializer. The compiler should prohibit this, however it doesn't.

Comment (r4zzz4k) [reply OBJC_INTEROP.md]: I see, so this is temporary state of thins. And so even if I want to use no-arg constructor without working with nibs it should still invoke `super(null, null)` to reference `UIViewController(nibNameOrNil: String?, bundle: NSBundle?)`. Is that correct? If so, something like this should be great to have for newcomers like me not to fail this trap:
> Note: When implementing constructors of Kotlin classes which are extending Objective-C ones make sure you are calling superclass constructor corresponding to it's [*designated initializer*](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/Initialization.html#//apple_ref/doc/uid/TP40014097-CH18-ID217) (i.e. [`initWithNibName:bundle:`](https://developer.apple.com/documentation/uikit/uiviewcontroller/1621359-initwithnibname) for `UIViewController`).

Comment (SvyatoslavScherbina) [reply OBJC_INTEROP.md]: > if I want to use no-arg constructor without working with nibs it should still invoke `super(null, null)` to reference `UIViewController(nibNameOrNil: String?, bundle: NSBundle?)`. 

Yes. Swift also has this requirement.

We will either add such a warning to this doc or forbid this case in the compiler.

File: OBJC_INTEROP.md (hunk 1 above)
Comment (olonho) [reply OBJC_INTEROP.md]: Imported or used is fine with me.

File: OBJC_INTEROP.md (hunk 11 above)
Comment (olonho) [reply OBJC_INTEROP.md]: Maybe some high level description of initBy and current state of things around initializers makes sense indeed.

File: OBJC_INTEROP.md [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+
+```
+func foo(block: (NSNumber) -> KotlinUnit)
+```
+
+and can be called like
+```
+foo {
+    bar($0 as! Int32)
+    return KotlinUnit()
```
Comment (olonho) [inline OBJC_INTEROP.md]: Not too pretty, but we could live with it.

File: OBJC_INTEROP.md [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+val nsNumber = 42 as NSNumber
+```
+
+## Subclassing
+
+Kotlin classes and interfaces can be subclassed by Swift/Objective-C classes
+and protocols.
+Currently a class that adopts Kotlin protocol should inherit `NSObject`
+(either directly or indirectly). Note that all Kotlin classes do inherit `NSObject`,
+so a subclass of Kotlin class can adopt Kotlin protocol.
```
Comment (olonho) [inline OBJC_INTEROP.md]: Not entirely clear to me. Maybe "subclass of Kotlin class" -> "Objective-C subclass of Kotlin class"?

Comment (SvyatoslavScherbina) [reply OBJC_INTEROP.md]: It makes sense.

File: OBJC_INTEROP.md (hunk 11 above)
Comment (SvyatoslavScherbina) [reply OBJC_INTEROP.md]: @olonho btw, I'm working on some improvements to Objective-C initializers support, which would make `initBy` (and also calling/overriding `init*` methods) obsolete. So I suggest waiting until this work is done and then completing the documentation about initializers.

Comment (SvyatoslavScherbina) [reply OBJC_INTEROP.md]: And here it is: #1512

**Pull request merged**